pkgs/applications/networking/browsers/chromium/update.py

```diff
@@ -0,0 +1,51 @@
+#! /usr/bin/env nix-shell
+#! nix-shell -i python -p python3 -p nix
+
+import csv
+import json
+import re
+import subprocess
+from multiprocessing import Pool
+from codecs import iterdecode
+from os.path import dirname
+from urllib.request import urlopen
+
+DEB_URL = "https://dl.google.com/linux/chrome/deb/pool/main/g"
+BUCKET_URL = "https://commondatastorage.googleapis.com/chromium-browser-official";
+
+def nix_prefetch_url(url, hash="sha256"):
+    out = subprocess.check_output(["nix-prefetch-url", "--type", hash, url])
```

Comment:
Addendum to that and to explain the reason for the `builtins.tryEval (builtins.fetchurl url)` hack:

The old updater not only checked the main omahaproxy URL but also  `http://omahaproxy.appspot.com/history` and whenever a given tarball doesn't exist in the main URL, the updater would walk back the history until it finds a version where the tarball actually exists. 

This is done only via a `curl -I` (HTTP HEAD) request and is used to get an overview of all the latest versions that *actually* exist in the upstream bucket.

Only the result of this is compared against the existing `upstream-info.nix`, so that the updater always produced a valid result of the latest *available* versions, even if some upstream tarballs are non-existing.

Reply:
Shouldn't `nix-prefetch-url` return an error if the url to be fetched shows a 404?

Other than that, yeah, the updater could be extended to also probe `http://omahaproxy.appspot.com/history` (and we'd need to add this to the `src` attribute too)

Reply:
> Shouldn't `nix-prefetch-url` return an error if the url to be fetched shows a 404?

Of course and that's why I introduced that hash collision hack to do `builtins.tryEval (builtins.fetchurl url)` natively in Nix.

> Other than that, yeah, the updater could be extended to also probe `http://omahaproxy.appspot.com/history` (and we'd need to add this to the `src` attribute too)

Especially since back in the days where I was maintaining Chromium, this happened **very** frequently (like on almost every single update). Not sure however if this is still the case, but from my comment above it still happened in November 2019.

Reply:
Yeah, I think we might want to have the update script figure out the real location where things can be downloaded from, and then just write that URL to `upstream-info.json`, removing all the complicated "look in many different places" logic at build time.